### PR TITLE
Adds virtualservice tests

### LINKED ISSUE #417: istio/virtualservice.go missing tests
/kind feature

virtualservice.go has a ton of logic but is missing unit tests (only has integration tests). Need to add tests that test for:
* Default only or default+canary
* Predictor only vs predictor + transformer
* With and without explainer
* Shape mismatch between default, canary
* Ready vs unready

Also for `getPredictStatusConfigurationSpec` consider the interface: (kfsvc, isCanary). It allows us to have access to the entire object which can be useful when wrapping metadata, etc. Otherwise we will need to keep refactoring this function signature over time (which has happened elsewhere).

### PR DESCRIPTION
**Pull request by rakelkar**
Fixes #417 

/assign @yuzisun @ellis-bigelow

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/465)

### CODE REVIEW

**Comment by rakelkar** [comment]
@yuzisun @hougangliu can take a look?

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go [hunk 1]
```diff
@@ -237,8 +244,14 @@ func getPredictStatusConfigurationSpec(endpointSpec *v1alpha2.EndpointSpec, comp
 		return nil, PredictorSpecMissing
 	}
 
+	if componentStatusMap == nil {
```
Comment (ellis-bigelow) [inline pkg/controller/inferenceservice/resources/istio/virtualservice.go]: Big ask but:
Should we call StatusConfigurationSpec a ComponentStatus? I feeel like StatusSpec is an oxymoron.

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go [hunk 2]
```diff
@@ -237,8 +244,14 @@ func getPredictStatusConfigurationSpec(endpointSpec *v1alpha2.EndpointSpec, comp
...
+		return nil, PredictorStatusUnknown
+	}
+
 	if predictorStatus, ok := (*componentStatusMap)[constants.Predictor]; !ok {
 		return nil, PredictorStatusUnknown
+	} else if predictorStatus == nil {
+		return nil, PredictorStatusUnknown
 	} else if len(predictorStatus.Hostname) == 0 {
 		return nil, PredictorHostnameUnknown
 	} else {
```
Comment (ellis-bigelow) [inline pkg/controller/inferenceservice/resources/istio/virtualservice.go]: for the "", would it be cleaner to return a string*?

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go [hunk 3]
```diff
@@ -251,8 +264,14 @@ func getTransformerStatusConfigurationSpec(endpointSpec *v1alpha2.EndpointSpec,
 		return nil, TransformerSpecMissing
 	}
 
+	if componentStatusMap == nil {
```
Comment (ellis-bigelow) [inline pkg/controller/inferenceservice/resources/istio/virtualservice.go]: Seeing all these checks for each compontent, I smell an opportunity to generalize. Workable?

File: pkg/controller/inferenceservice/resources/istio/virtualservice_test.go [hunk 4]
```diff
@@ -0,0 +1,466 @@
+/*
```
Comment (ellis-bigelow) [inline pkg/controller/inferenceservice/resources/istio/virtualservice_test.go]: You are a saint.

File: pkg/controller/inferenceservice/resources/istio/virtualservice_test.go [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+		defaultStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: &v1alpha2.StatusConfigurationSpec{
+				Hostname: predictorHostname,
+			},
+		},
+		canaryStatus: &v1alpha2.ComponentStatusMap{
+			constants.Predictor: &v1alpha2.StatusConfigurationSpec{},
+		},
+		expectedStatus:  createFailedStatus(PredictorHostnameUnknown, PredictorMissingMessage),
+		expectedService: nil,
```
Comment (ellis-bigelow) [inline pkg/controller/inferenceservice/resources/istio/virtualservice_test.go]: I've seen a convention where you leave out nil values from the map and let it default.

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go (hunk 1 above)
Comment (rakelkar) [reply pkg/controller/inferenceservice/resources/istio/virtualservice.go]: seems like can be a separate change?

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go (hunk 3 above)
Comment (rakelkar) [reply pkg/controller/inferenceservice/resources/istio/virtualservice.go]: I think the way to do that is to have base component that allows us to treat this as a DAG. Out of scope of this humble change though.

File: pkg/controller/inferenceservice/resources/istio/virtualservice_test.go (hunk 5 above)
Comment (rakelkar) [reply pkg/controller/inferenceservice/resources/istio/virtualservice_test.go]: um - nice to have it explicit in this case?

File: pkg/controller/inferenceservice/resources/istio/virtualservice_test.go [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+			}
+
+			if diff := cmp.Diff(tc.expectedService, actualService); diff != "" {
+				t.Errorf("Test %q unexpected service (-want +got): %v", tc.name, diff)
+			}
+		})
+	}
+}
+
+func createMockPredictorSpec(componentStatusMap *v1alpha2.ComponentStatusMap) v1alpha2.PredictorSpec {
```
[the remark this answers is not in the document]
Comment (rakelkar) [reply pkg/controller/inferenceservice/resources/istio/virtualservice_test.go]: Just a way of setting up the actualService (based on actualStatus) to minimize the test setup

**Pull request merged**